src/source-windivert.c

```diff
@@ -0,0 +1,891 @@
+static void WinDivertRestoreOffloading(WinDivertThreadVars *);
+
+void TmModuleReceiveWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_RECEIVEWINDIVERT];
+
+    tm_ptr->name = "ReceiveWinDivert";
+    tm_ptr->ThreadInit = ReceiveWinDivertThreadInit;
+    tm_ptr->PktAcqLoop = ReceiveWinDivertLoop;
+    tm_ptr->ThreadExitPrintStats = ReceiveWinDivertThreadExitStats;
+    tm_ptr->ThreadDeinit = ReceiveWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_RECEIVE_TM;
+}
+
+void TmModuleVerdictWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_VERDICTWINDIVERT];
+
+    tm_ptr->name = "VerdictWinDivert";
+    tm_ptr->ThreadInit = VerdictWinDivertThreadInit;
+    tm_ptr->Func = VerdictWinDivert;
+    tm_ptr->ThreadDeinit = VerdictWinDivertThreadDeinit;
+}
+
+void TmModuleDecodeWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_DECODEWINDIVERT];
+
+    tm_ptr->name = "DecodeWinDivert";
+    tm_ptr->ThreadInit = DecodeWinDivertThreadInit;
+    tm_ptr->Func = DecodeWinDivert;
+    tm_ptr->ThreadDeinit = DecodeWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_DECODE_TM;
+}
+
+/**
+ * \brief Main WinDivert packet receive pump
+ */
+TmEcode ReceiveWinDivertLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    WinDivertThreadVars *wd_tv = (WinDivertThreadVars *)data;
+    wd_tv->slot = ((TmSlot *)slot)->slot_next;
+
+    while (true) {
+        if (suricata_ctl_flags & SURICATA_STOP) {
+            SCReturnInt(TM_ECODE_OK);
+        }
+
+        if (unlikely(WinDivertRecvHelper(tv, wd_tv) != TM_ECODE_OK)) {
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+
+        StatsSyncCountersIfSignalled(tv);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode WinDivertRecvHelper(ThreadVars *tv, WinDivertThreadVars *wd_tv)
+{
+    SCEnter();
+
+#ifdef COUNTERS
+    WinDivertQueueVars *wd_qv = WinDivertGetQueue(wd_tv->thread_num);
+#endif /* COUNTERS */
+
+    /* make sure we have at least one packet in the packet pool, to prevent us
+     * from alloc'ing packets at line rate
+     */
+    PacketPoolWait();
+
+    /* obtain a packet buffer */
+    Packet *p = PacketGetFromQueueOrAlloc();
+    if (unlikely(p == NULL)) {
+        SCLogDebug(
+                "PacketGetFromQueueOrAlloc() - failed to obtain Packet buffer");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+
+    /* receive packet, depending on offload status. MTU is used as an estimator
+     * for direct data alloc size, and this is meaningless if large segments are
+     * coalesced before they reach WinDivert */
+    bool success = false;
+    if (wd_tv->offload_enabled) {
+        /* allocate external, if not already */
+        PacketCallocExtPkt(p, MAX_PAYLOAD_SIZE);
+
+        success = WinDivertRecv(wd_tv->filter_handle, p->ext_pkt,
+                                MAX_PAYLOAD_SIZE, &p->windivert_v.addr,
+                                &p->pktlen);
```

Comment:
our coccinelle check for pkt access triggers on this:
```
Using concurrency level 28
Testing cocci file: access-pkt-packet.cocci
Invalid Packet->pktlen usage, GET_PKT_LEN macro must be used at /builds/inliniac/suricata-ci/suricata/qa/coccinelle/../../src/source-windivert.c:342
ERROR: Job failed: exit code 1
```

Reply:
I guess the way to work around this would be to use a local var to get the pktlen from WinDivertRecv and then use the SET_PKT_LEN macro.